src/JsonPointer/JsonPointer.cs

```diff
@@ -69,64 +67,183 @@ public class JsonPointer : IEquatable<JsonPointer>, IReadOnlyList<string>
-	public static JsonPointer Parse(string source)
+	public static JsonPointer Parse(ReadOnlySpan<char> source)
 	{
 		if (source.Length == 0) return Empty;
 
 		if (source[0] == '#')
-			source = HttpUtility.UrlDecode(source[1..]);  // allocation
+		{
+#if NET9_0_OR_GREATER
+			source = Uri.UnescapeDataString(source[1..]);
+#elif NET8_0_OR_GREATER
+			source = Uri.UnescapeDataString(new (source[1..]));
+#else
+			source = Uri.UnescapeDataString(source[1..].ToString()).AsSpan();
+#endif
+
+			if (source.Length == 0) return Empty;
+		}
 
-		if (source.Length == 0) return Empty;
 		if (source[0] != '/')
 			throw new PointerParseException("Pointer must start with either `#/` or `/` or be empty");
 
+		if (source.Length == 1)
+			return new JsonPointer([""]);
+
+		if (TryParseInternal(source, out var result))
+			return result;
+
+		throw new PointerParseException($"Value '{source.ToString()}' does not represent a valid JSON Pointer");
+	}
+
+	private static int CountSegments(ReadOnlySpan<char> source)
+	{
+		var size = 0;
+		var pos = source.IndexOf('/');
+
+		while (pos >= 0)
+		{
+			size++;
+
+			source = source[(pos + 1)..];
+			pos = source.IndexOf('/');
+		}
+
+		return size;
+	}
+
+#if NET9_0_OR_GREATER
+
+	private static bool TryParseInternal(ReadOnlySpan<char> source, [NotNullWhen(true)] out JsonPointer? result)
+	{
+		var segmentCount = CountSegments(source);
+		var segments = new string[segmentCount];
+		var segmentIndex = 0;
+
+		source = source[1..];
+		foreach (var segmentRange in source.Split('/'))
+		{			
+			if (!TryDecodeSegment(source[segmentRange], out segments[segmentIndex++]))
+			{
+				result = null;
+				return false;
+			}	
+		}
+
+		result = new JsonPointer(segments);
+		return true;
+	}
+
+#else
+
+	private static bool TryParseInternal(ReadOnlySpan<char> source, [NotNullWhen(true)] out JsonPointer? result)
+	{
+		var segmentCount = CountSegments(source);
+		var segments = new string[segmentCount];
+
+		source = source[1..];
 		var segmentIndex = 0;
-		var sourceIndex = 1;
-		var builderIndex = 0;
-		var sourceSpan = source.AsSpan();
-		using var segmentMemory = MemoryPool<string>.Shared.Rent();
-		var segments = segmentMemory.Memory.Span;
-		using var builderMemory = MemoryPool<char>.Shared.Rent();
-		var builder = builderMemory.Memory.Span;
-		while (sourceIndex < source.Length)
+		result = null;
+
+		var sourceIndex = source.IndexOf('/');
+		while (sourceIndex >= 0)
 		{
-			if (source[sourceIndex] == '/')
+			if (!TryDecodeSegment(source[..sourceIndex], out var segment))
+				return false;
+
+			segments[segmentIndex++] = segment;
+			
+			source = source[(sourceIndex + 1)..];
+			sourceIndex = source.IndexOf('/');
+		}
+
+		if (!TryDecodeSegment(source, out segments[segmentIndex]))
+			return false;
+		
+		result = new JsonPointer(segments);
+		return true;
+	}
+
+#endif
+
+	private static bool TryDecodeSegment(ReadOnlySpan<char> encoded, out string? result)
```

Comment:
Please add nullability annotation, and please put back into the `SpanExtensions` class.  I think it reads better that way.